omnimove_external_control/include/omnimove/external_control_message.hpp

```diff
@@ -0,0 +1,145 @@
+#ifndef EXTERNAL_CONTROL_MESSAGE_H
+#define EXTERNAL_CONTROL_MESSAGE_H
+#include <boost/asio.hpp>
+#include <cstdint>
+#include <memory>
+
+namespace omnimove
+{
+
+class ExternalControlMessage
+{
+private:
+  std::unique_ptr<char[]> msg_buffer_;
+  size_t msg_length_;
+
+protected:
+  const char * header_;
+  const char * message_;
+  const char * crc_;
+  ExternalControlMessage(const char * header, const char * message, const int & msg_length);
+
+public:
+  int16_t calculateCRC();
+  boost::asio::const_buffer getSerialisedData();
+};
+
+class ExternalControlData : public ExternalControlMessage
+{
+private:
+  const bool is_valid_;
+
+protected:
+  static constexpr const char * EXTERNAL_CONTROL_DATA_HEADER = "KMRUTV03";
+  static constexpr int EXTERNAL_CONTROL_DATA_LENGTH = 106;  // without the header and crc bytes
+  uint8_t actual_mode_;
+  int32_t actual_speed_x_;
+  int32_t actual_speed_y_;
+  int32_t actual_speed_w_;
+  uint8_t agv_is_moving_;
+  int32_t max_agv_speed_x_;
+  int32_t max_agv_speed_y_;
+  int32_t max_agv_speed_w_;
+  int32_t deviation_x_;
+  int32_t deviation_y_;
+  int32_t deviation_w_;
+  uint8_t dev_driving_;
+  uint8_t dev_target_reached_;
+  uint8_t scanning_fields_;
+  uint8_t driving_direction_;
+  uint8_t current_state_;
+  uint32_t dataMatrix_FrontCamera;
+  uint32_t dataMatrix_RearCamera_;
+  uint32_t dataMatrix_LeftCamera_;
+  uint32_t dataMatrix_RightCamera;
+  uint32_t status_optical_track_guiding_;
+  uint8_t warning_optical_track_guiding_;
+  uint8_t error_optical_track_guiding_;
+  int32_t track_deviation_x_;
+  int32_t track_deviation_y_;
+  int32_t track_deviation_w_;
+  uint8_t tag_number_front_camera_;
+  uint8_t tag_number_rear_camera_;
+  uint8_t tag_number_left_camera_;
+  uint8_t tag_number_right_camera_;
+  struct PillarData
+  {
+    uint8_t manual_active_;
+    uint8_t target_height_reached_;
+    uint8_t actual_speed_;
+    uint8_t actual_pos_;
+  };
+
+  PillarData pillar1_;
+  PillarData pillar2_;
+  PillarData pillar3_;
+  PillarData pillar4_;
```

Comment:
the four pillars could be combined to a single std::array (or std::vector), which would also make copying easier and shorter in some cases

Reply:
done. combined the four pillars into one